File: sln/70. Climbing Stairs_v1.cpp
/*70. Climbing Stairs
You are climbing a stair case. It takes n steps to reach to the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

Note: Given n will be a positive integer.

Example 1:

Input: 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps
Example 2:

Input: 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step

*/
class Solution {
private:
    vector<int> memo; 
    
    int calWays(int n) {
        if (n == 0 || n == 1) {
            return 1;
        }
        if (memo[n] == -1)
            memo[n] = calWays(n-1) + calWays(n-2);
        return memo[n];
    }
public:
    int climbStairs(int n) {
        memo = vector<int>(n+1, -1);
        return calWays(n);
    }
};

File: sln/70. Climbing Stairs_v1.java
class Solution {
    
    public int climbStairs(int n) {
        
        res = new int[n+1];
        return f(n);              
    }
    // f(1) = 1
    // f(2) = 1
    // f(n) = f(n-1) + f(n-2);
    // if we climb 1 step firstly, we have f(n-1) ways to climb later;
    // if we climb 2 steps firstly, we have f(n-2) ways to climber later;
    private int[] res;
    
    private int f(int n) {
        
        if (res[n] != 0) return res[n]; 
        
        if (n == 1 || n == 2) res[n] = n;
        else res[n] =  f(n-1) + f(n-2);
        
        return res[n];
    }
}